webui/react/src/routes/index.ts

```diff
@@ -12,173 +11,38 @@ import TaskList from 'pages/TaskList';
 import TaskLogs from 'pages/TaskLogs';
 import TrialDetails from 'pages/TrialDetails';
 import TrialLogs from 'pages/TrialLogs';
-import history from 'routes/history';
-import { clone } from 'utils/data';
-import { isFullPath, parseUrl } from 'utils/routes';
 
-/*
- * Router Configuration
- * If the component is not defined, the route is assumed to be an external route,
- * meaning React will attempt to load the path outside of the internal routing
- * mechanism.
- */
-export interface RouteConfig extends RouteProps {
-  id: string;
-  icon?: string;
-  path: string;
-  popout?: boolean;
-  redirect?: string;
-  suffixIcon?: string;
-  title?: string;
-  needAuth?: boolean;
-}
+import Routes from './routes';
+import { RouteConfig } from './types';
 
-const dashboardRoute = {
-  component: Dashboard,
-  icon: 'user',
-  id: 'dashboard',
-  needAuth: true,
-  path: '/det/dashboard',
-  title: 'Dashboard',
+const routeComponentMap: Record<string, FC> = {
+  cluster: Cluster,
+  dashboard: Dashboard,
+  experimentDetails: ExperimentDetails,
+  experimentList: ExperimentList,
+  masterLogs: MasterLogs,
+  signIn: SignIn,
+  signOut: SignOut,
+  taskList: TaskList,
+  taskLogs: TaskLogs,
+  trialDetails: TrialDetails,
+  trialLogs: TrialLogs,
 };
 
-const clusterRoute = {
-  component: Cluster,
-  icon: 'cluster',
-  id: 'cluster-det',
-  needAuth: true,
-  path: '/det/cluster',
-  title: 'Cluster',
-};
-
-const experimentListRoute = {
-  component: ExperimentList,
-  icon: 'experiment',
-  id: 'experimentList',
-  needAuth: true,
-  path: '/det/experiments',
-  title: 'Experiments',
-};
-
-const experimentDetailsRoute = {
-  component: ExperimentDetails,
-  id: 'experimentDetails',
-  needAuth: true,
-  path: '/det/experiments/:experimentId',
-  title: 'Experiment',
-};
-
-const trialDetailsRoute = {
-  component: TrialDetails,
-  id: 'trialDetails',
-  needAuth: true,
-  path: '/det/trials/:trialId',
-  title: 'Trial',
-};
-
-const taskLogsRoute = {
-  component: TaskLogs,
-  icon: 'logs',
-  id: 'task-logs',
-  needAuth: true,
-  path: '/det/:taskType/:taskId/logs',
-  title: 'Task Logs',
-};
+const defaultRouteId = 'dashboard';
+const defaultRoute = Routes.find(route => route.id === defaultRouteId);
 
-const taskListRoute = {
-  component: TaskList,
-  icon: 'tasks',
-  id: 'taskList',
-  needAuth: true,
-  path: '/det/tasks',
-  title: 'Tasks',
-};
+export const defaultRoutePath = defaultRoute?.path || '/det/dashboard';
 
-const masterLogsRoute = {
-  component: MasterLogs,
-  icon: 'logs',
-  id: 'logs',
-  needAuth: true,
-  path: '/det/logs',
-  title: 'Master Logs',
-};
+const appRoutes: RouteConfig[] = Routes.map(route => ({
+  ...route,
+  component: routeComponentMap[route.id],
```

Comment:
this could be undefined because theoretically route.id could be missing. 
can we use keyof to  safeguard against what can be set in route.id? 

overall I think the previous setup where the route and component where defined together makes more sense and if that means we have to define a new file for some of our utilities to avoid circular dependencies  or having to move `isReactRoute` into this module that's fine.

Reply:
the `keyof typeof routeComponentMap` casting doesn't seem to provide actual checking.
```
  component: routeComponentMap[route.id as keyof typeof routeComponentMap],
```
The alternative would be to extract all the `route.id`s first then create a type out of that and setting that to the key type of `Record<whatever-type, RouteConfig>`. Instead decided to throw an exception if the `route.id` doesn't exist in routeComponentMap. It's serious enough of an issue that it should halt the webapp.

Reply:
would it provide type checking if we don't explicitly define routeComponentMaps type? (although that might have its own downsides but probably still safer) I haven't checked if this would get us into that circular dependency again or not.

![Screen Capture_select-area_20200914110517](https://user-images.githubusercontent.com/12127420/93121739-3262cc00-f67a-11ea-878d-7cfc9b3f34ff.png)

Reply:
If we drop the `Record<string, FC>` definition on `routeComponentMap`, then this line does signal an error:
```
component: routeComponentMap[router.id],
```
because it's treating `router.id` as a `string` type. So when we recast it to `keyof typeof routeComponentMap` it fixes the complaint but even if an invalid `route.id` is added into `Routes`, TypeScript doesn't complain.